### PR TITLE
--ini error conditions

### PR DESCRIPTION
**Pull request by samstav**
This change raises exceptions on these conditions:
    - an explicitly defined ini file does not exist
    - a config option found in a loaded ini file
      does not correspond to a `config.Option`

Add `SimplConfigException` and `SimplConfigUnknownOption`

`SimplConfigUnknownOption` will be thrown if an ini file is
explicitly specified, but there are options in it that have
no match, i.e. no corresponding `config.Option`

Previously, if these conditions occurred, nothing would happen. @Gifflen spent quite a long time trying to figure out why the options defined in his ini file were not getting picked up by simpl.

### CODE REVIEW

File: simpl/exceptions.py [hunk 1]
```diff
@@ -107,3 +107,13 @@ def __str__(self):
...
+
+
+class SimplConfigException(SimplException):
+
+    """Errors raised by simpl/config."""
+
+
+class SimplConfigUnusedIniOption(SimplConfigException):
+
+    """An option defined in the specified ini file has no match."""
```
Comment (samstav) [inline simpl/exceptions.py]: self-critique... This exception name is bad.

File: simpl/config.py [hunk 2]
```diff
@@ -526,7 +525,7 @@ def get_defaults(self):
         parsed, _ = parser.parse_known_args([])
         return vars(parsed)
 
-    def parse_ini(self, paths=None, namespace=None):
+    def parse_ini(self, paths=None, namespace=None, permissive=False):
```
Comment (samstav) [inline simpl/config.py]: I added the permissive argument, but there's not really any easy way to use it, unless you were to subclass Config and override the `load_options` method. 

I assumed (and agree) that throwing an error upon encountering "loose" options might not be the desired behavior for some. For the record, this works very well for our use cases, and would have alerted us to problems sooner, although I would argue that it would be nice if you could turn this feature on more easily, i.e. the `permissive` feature of our ini parsing.

File: simpl/exceptions.py (hunk 1 above)
Comment (ziadsawalha) [inline simpl/exceptions.py]:
```
$python -f
Unknown option: -f
```

How about `SimplConfigUnknownOption` and we use it for arguments as well as ini options?

File: simpl/config.py (hunk 2 above)
Comment (ziadsawalha) [inline simpl/config.py]: Supporting different parsing between ini files and command arguments will make the interface more complicated than needed. So I think if we add it we should add `permissive=` kwarg (instead of `permissive_ini=, permissive_args, etc..`
But... having said that... we can leave this out until it is required. As you say, we don't need it for our use cases, so we could skip it until (if ever) someone asks for it.
In other words, the code as is is good, IMO

File: simpl/config.py [hunk 3]
```diff
@@ -165,10 +165,13 @@ def main(argv):
 import argparse
 import collections
 import copy
+import errno
 import logging
 import os
 import sys
 
+from simpl import exceptions as simpl_exceptions
```
Comment (ziadsawalha) [inline simpl/config.py]: It would be nice if we came up with a convention for importing exceptions.

Some things we've discussed before that maybe we can add to a contributing.md or in the docstring for simpl/exceptions.py:
- A fully-named exception is useful when exceptions are caught and bubble up (ex. SimplConfigError is easier to track than simpl.ConfigError if it bubbles out of simpl)
- "exceptions" is a long word. One of our projects puts exceptions in `exc.py`; I think that's a good idea.
- valid reasons for renaming imports are to avoid conflicts or to shorten very long names.
simple_exceptions is no shorter than simpl.exceptions. So why not just:
```python
import simple.exceptions

...
raise simpl.exceptions.SimplConfigUnusedIniOption()
```

File: tests/test_config.py [hunk 4]
```diff
@@ -42,6 +44,22 @@ def test_format_comma_separated_pairs(self):
 
 
 class TestConfig(unittest.TestCase):
+
+    def setUp(self):
```
Comment (ziadsawalha) [inline tests/test_config.py]: Should we PEP test files? Maybe not full analysis, but at least PEP-8?

File: tests/test_git.py [hunk 5]
```diff
@@ -283,15 +283,25 @@ def test_clone_brings_branches(self):
...
-        self.assertIn({
-            'commit': self.repo.head,
-            'message': 'On master: stash message',
-            'branch': '(detached from %s)' % self.repo.head[:7]
-        }, branches)
+        selected = None
+        for b in branches:
+            if b['commit'] == self.repo.head:
+                selected = b
+                break
```
Comment (ziadsawalha) [inline tests/test_git.py]: No changes required, just musings on python...

I wonder/wish if there was a simpler pattern for this in python. Its such a common occurrence.

This works but its not inconclusively simpler or clearer.
```python
        selected = next((b for b in branches if b['commit'] == self.repo.head), None)
```

File: tests/test_git.py [hunk 6]
```diff
@@ -283,15 +283,25 @@ def test_clone_brings_branches(self):
...
+        self.assertIsNotNone(selected)
+        self.assertEqual(selected['commit'], self.repo.head)
+        self.assertEqual(selected['message'], 'On master: stash message')
+        try:
+            fmtr = {'cmt': self.repo.head[:7]}
+            self.assertEqual(
+                selected['branch'], '(detached from {cmt})'.format(**fmtr))
+        except AssertionError as e:
```
Comment (ziadsawalha) [inline tests/test_git.py]: Are these different possible responses based on the version of git?

File: simpl/config.py [hunk 7]
```diff
@@ -538,13 +537,24 @@ def parse_ini(self, paths=None, namespace=None):
...
+        for pth in inipaths:
+            if not os.path.isfile(pth):
+                raise OSError(errno.ENOENT, 'No such file or directory', pth)
+
+        read_ok = self.ini_config.read(inipaths)
+        assert read_ok == inipaths
+        dicts = (list(self.ini_config._sections.values()) +
+                 [self.ini_config.defaults()])
+        ini_options = copy.deepcopy(
+            {k: v for d in dicts for k, v in d.items() if k != '__name__'})
```
Comment (samstav) [inline simpl/config.py]: Change this to a `set` of keys rather than a dict. You don't need the values.

File: simpl/config.py (hunk 2 above)
Comment (samstav) [inline simpl/config.py]: :+1: Thanks.

File: tests/test_git.py (hunk 5 above)
Comment (samstav) [inline tests/test_git.py]: > If they had called it `elifnobreak`, nobody would be confused.

- Raymond Hettinger

-----------

https://youtu.be/OSGv2VnC0go

File: tests/test_git.py (hunk 6 above)
Comment (samstav) [inline tests/test_git.py]: Yes, I'll add a comment.

Comment (samstav) [inline tests/test_git.py]: https://github.com/samstav/simpl/commit/03068c2da07ce5b26306d8922883fb2e7e49bd7e

**Comment by samstav** [comment]
Thanks for the thorough review @ziadsawalha. I think I've addressed all comments, let me know.

File: simpl/config.py [hunk 8]
```diff
@@ -165,10 +165,13 @@ def main(argv):
 import argparse
 import collections
 import copy
+import errno
 import logging
 import os
 import sys
 
+import simpl.exceptions
```
Comment (ziadsawalha) [inline simpl/config.py]: Much cleaner, IMO. Thanks.

File: tests/test_config.py [hunk 9]
```diff
@@ -42,6 +44,13 @@ def test_format_comma_separated_pairs(self):
 
 
 class TestConfig(unittest.TestCase):
+
+    def get_tempfile(self, *args, **kwargs):
+
```
Comment (ziadsawalha) [inline tests/test_config.py]: Blank line after `def` contradicts PEP-8.

**Comment by ziadsawalha** [comment]
:+1: - last two comments are minor.
